x/stake/msg.go

```diff
@@ -0,0 +1,227 @@
+package stake
+
+import (
+	"encoding/json"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	crypto "github.com/tendermint/go-crypto"
+)
+
+// name to idetify transaction types
+const MsgType = "stake"
+
+// XXX remove: think it makes more sense belonging with the Params so we can
+// initialize at genesis - to allow for the same tests we should should make
+// the ValidateBasic() function a return from an initializable function
+// ValidateBasic(bondDenom string) function
+const StakingToken = "fermion"
+
+//Verify interface at compile time
+var _, _, _, _ sdk.Msg = &MsgDeclareCandidacy{}, &MsgEditCandidacy{}, &MsgDelegate{}, &MsgUnbond{}
+
+//______________________________________________________________________
+
+// MsgDeclareCandidacy - struct for unbonding transactions
+type MsgDeclareCandidacy struct {
+	Description
+	CandidateAddr sdk.Address   `json:"address"`
+	PubKey        crypto.PubKey `json:"pubkey"`
+	Bond          sdk.Coin      `json:"bond"`
+}
+
+func NewMsgDeclareCandidacy(candidateAddr sdk.Address, pubkey crypto.PubKey,
+	bond sdk.Coin, description Description) MsgDeclareCandidacy {
+	return MsgDeclareCandidacy{
+		Description:   description,
+		CandidateAddr: candidateAddr,
+		PubKey:        pubkey,
+		Bond:          bond,
+	}
+}
+
+//nolint
+func (msg MsgDeclareCandidacy) Type() string                            { return MsgType } //TODO update "stake/declarecandidacy"
+func (msg MsgDeclareCandidacy) Get(key interface{}) (value interface{}) { return nil }
+func (msg MsgDeclareCandidacy) GetSigners() []sdk.Address               { return []sdk.Address{msg.CandidateAddr} }
+func (msg MsgDeclareCandidacy) String() string {
+	return fmt.Sprintf("CandidateAddr{Address: %v}", msg.CandidateAddr) // XXX fix
+}
+
+// get the bytes for the message signer to sign on
+func (msg MsgDeclareCandidacy) GetSignBytes() []byte {
+	b, err := json.Marshal(msg)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+// quick validity check
+func (msg MsgDeclareCandidacy) ValidateBasic() sdk.Error {
+	if msg.CandidateAddr == nil {
+		return ErrCandidateEmpty()
+	}
+	if msg.Bond.Denom != StakingToken {
+		return ErrBadBondingDenom()
+	}
+	if msg.Bond.Amount <= 0 {
+		return ErrBadBondingAmount()
+		// return sdk.ErrInvalidCoins(sdk.Coins{msg.Bond}.String())
+	}
+	empty := Description{}
+	if msg.Description == empty {
+		return newError(CodeInvalidInput, "description must be included")
+	}
+	return nil
+}
+
+//______________________________________________________________________
+
+// MsgEditCandidacy - struct for editing a candidate
+type MsgEditCandidacy struct {
+	Description
+	CandidateAddr sdk.Address `json:"address"`
+}
```

Comment:
Spec also includes provision for changing`commission`

Reply:
commission not yet implemented